### ISSUE TITLE
Overriding gradients in eager mode with TF 2 after the model is defined

### ISSUE THREAD

**Issue by slundberg** [tensorflow/tensorflow#35203]
When I wrote SHAP for TensorFlow 1.x I was able to temporarily override the gradient operators inside `tensorflow.python.framework.ops._gradient_registry._registry`. This allowed me to use the back-propagation mechanisms of TF to compute explainable AI methods on models that had already been defined by the user. I couldn't use the `gradient_override_map` since that would require the user to build their model within my SHAP context...and I need to assume they already have a model they have built. Since I can "back-propagate" explanations through non-differentiable functions I also need to temporarily override `tensorflow.python.ops.gradients_impl._IsBackpropagatable` so my gradient handlers get called everywhere and things don't get pruned from the graph. Both of these are not public APIs, but there was no other way I found to get it to work.

Now in TF 2.0 I can't seem to replicate this while in eager mode. It looks like I can get everything to work using `tf.GradientTape` when everything between the input and the output is differentiable. But as soon as something that is not differentiable shows up (like ResourceGather) then it seems that in eager mode overriding `_IsBackpropagatable` no longer ensures that the gradient handlers will get called. I assume this is because the GradientTape doesn't think the "watched" tensor is connected to the output when the path through the graph includes non-differentiable operators. One would think that overriding  `_IsBackpropagatable` would fix this, but apparently not in TF 2.0 with eager mode. In fact I never see _IsBackpropagatable getting called at all in TF 2 eager mode. I was trying to figure out what the equivalent function is I can override, but with TF2 the GradientTape does all the primary work now in C++, so it is not as easy to trace through and get at what I need.

Any advice? Thanks!

**Comment by slundberg** [comment]
Quick update: It looks like the equivalent of `_IsBackpropagatable` in eager mode is now a hard-coded C++ type check on the op inputs at: https://github.com/tensorflow/tensorflow/blob/7b094e93b906f32b971b13c6b7fb5df75cde5cf6/tensorflow/c/eager/tape.h#L340-L369

**Comment by slundberg** [comment]
I finally worked around this by overriding `tensorflow.python.eager.execute.record_gradient` and temporarily lying about the index input type of `ResourceGather` making it a float instead of an int just during the duration of the C++ call `TFE_Py_RecordGradient`. This was the only way I could figure out to force the GradientTape to work with non-differential operators.

So while now things work, I am still very interested in suggestions to find a less hacky and hence more stable solution.

**Comment by lerobitaille** [comment]
Hi, I'm trying to implement defined in [guided backprop](https://arxiv.org/abs/1412.6806) using tensorflow 2.0. For this, I need to everride the gradient map to make ReLU layers rectify the gradient, example [here](https://stackoverflow.com/questions/38340791/guided-back-propagation-in-tensorflow). 

I've search for some time now and the only solution without hacking the code seems to swap to graph mode, but this is not ideal since it removes all the other 2.0 capabilities that I was using. Is there an other way yet? Thanks in advance!

**Comment by bmiselis** [comment]
I have a similar use case as @lerobitaille - I'd like to be able to override ReLU's default gradient calculation with a custom one. Previously it was possible to do it using `gradient_override_map`. Is there a way to replace gradient (even during the model loading is fine, too) in the eager execution of TensorFlow 2.0?

**Comment by alextp** [comment]
To be able to override gradients use tf.custom_gradient.

We don't really support overriding these things after the fact, though, just before the model is created.
